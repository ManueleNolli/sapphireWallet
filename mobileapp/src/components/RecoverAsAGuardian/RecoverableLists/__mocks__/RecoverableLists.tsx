const RecoverableListsMock = () => <div data-testid="RecoverableLists">RecoverableLists</div>

export default RecoverableListsMock
