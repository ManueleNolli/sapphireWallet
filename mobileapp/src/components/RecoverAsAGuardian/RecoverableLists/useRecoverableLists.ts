import { useContext, useEffect, useState } from 'react'
import useLoading from '../../../hooks/useLoading'
import { getGuardians, getGuardianWallets } from '../../../services/transactions'
import { WalletContext } from '../../../context/WalletContext'
import { BlockchainContext } from '../../../context/BlockchainContext'

export function useRecoverableLists() {
  const { getWalletContractAddress } = useContext(WalletContext)
  const { currentNetwork, ethersProvider } = useContext(BlockchainContext)
  const [wallets, setWallets] = useState<string[]>([])
  const { isLoading: isFetching, setIsLoading: setIsFetching } = useLoading(true)
  const [isRecovering, setIsRecovering] = useState(false)
  const [recoveringWallet, setRecoveringWallet] = useState<string>('')

  useEffect(() => {
    const fetchWallets = async () => {
      setWallets(await getGuardianWallets(ethersProvider, currentNetwork, getWalletContractAddress()))
    }
    fetchWallets()
    setIsFetching(false)
  }, [])

  const startRecovering = (wallet: string) => {
    setIsRecovering(true)
    setRecoveringWallet(wallet)
  }

  return {
    wallets,
    isFetching,
    isRecovering,
    startRecovering,
    recoveringWallet,
    modalBackdrop: () => setIsRecovering(false),
  }
}
