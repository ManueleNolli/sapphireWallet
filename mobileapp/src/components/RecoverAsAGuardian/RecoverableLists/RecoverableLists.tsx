import React from 'react'
import { View } from 'react-native'
import useRecoverableLists from './useRecoverableLists'
import { appStyles, vh, vw } from '../../../Styles'
import { Button, Modal, Spinner, StyleService, useStyleSheet, Text } from '@ui-kitten/components'
import { formatBlockchainAddress } from '../../../utils/formatBlockchainData'
import RecoverWallet from '../RecoverWallet/RecoverWallet'

type RecoverableListsProps = {
  refreshRequest: boolean
}

export default function RecoverableLists(props: RecoverableListsProps) {
  const { wallets, isFetching, isRecovering, startRecovering, recoveringWallet, modalBackdrop } =
    useRecoverableLists(props)
  const styles = useStyleSheet(themedStyles)

  const renderWallet = (wallet: string) => {
    return (
      <Button
        key={wallet}
        style={{ width: '100%', marginTop: 1 * vh }}
        appearance="outline"
        onPress={() => startRecovering(wallet)}
      >
        <Text>{formatBlockchainAddress(wallet)}</Text>
      </Button>
    )
  }

  return (
    <View>
      {isFetching && (
        <View style={appStyles.center}>
          <Spinner
            size="giant"
            status="basic"
            style={{
              marginBottom: 2 * vh,
            }}
          />
        </View>
      )}
      <Modal
        animationType="fade"
        visible={isRecovering}
        backdropStyle={styles.modalBackdrop}
        onBackdropPress={modalBackdrop}
      >
        <RecoverWallet wallet={recoveringWallet} />
      </Modal>

      {wallets.map(renderWallet)}
      {!isFetching && wallets.length === 0 && (
        <Text style={{ textAlign: 'center' }}>You are not a guardian of any wallet</Text>
      )}
    </View>
  )
}

const themedStyles = StyleService.create({
  modalContainer: {
    maxHeight: 75 * vh,
    maxWidth: 90 * vw,
    borderRadius: 10,
    paddingVertical: 4 * vh,
    paddingHorizontal: 4 * vw,
    alignItems: 'center',
  },
  modalBackdrop: {
    backgroundColor: 'color-basic-transparent-600',
  },
})
