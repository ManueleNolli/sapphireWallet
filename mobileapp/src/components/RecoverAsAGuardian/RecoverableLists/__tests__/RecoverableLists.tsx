import React from 'react'
import useRecoverableLists from '../useRecoverableLists'
import renderWithTheme from '../../../../TestHelper'
import RecoverableLists from '../RecoverableLists'
import { fireEvent } from '@testing-library/react-native'

jest.mock('../useRecoverableLists', () => jest.fn())
jest.mock('../../RecoverWallet/RecoverWallet')
jest.mock('@ui-kitten/components', () => {
  const { View } = require('react-native')
  return {
    ...jest.requireActual('@ui-kitten/components'),
    Spinner: () => <View data-testid="spinner" />,
  }
})
describe('RecoverableLists', () => {
  it('should render correctly when fetching', () => {
    ;(useRecoverableLists as jest.Mock).mockReturnValue({
      wallets: [],
      isFetching: true,
      isRecovering: false,
      startRecovering: jest.fn(),
      recoveringWallet: '',
      modalBackdrop: jest.fn(),
    })
    const tree = renderWithTheme(<RecoverableLists refreshRequest={false} />)

    expect(tree).toMatchSnapshot()
  })

  it('should render correctly when recovering', () => {
    ;(useRecoverableLists as jest.Mock).mockReturnValue({
      wallets: ['wallet1', 'wallet2'],
      isFetching: false,
      isRecovering: true,
      startRecovering: jest.fn(),
      recoveringWallet: 'wallet1',
      modalBackdrop: jest.fn(),
    })
    const tree = renderWithTheme(<RecoverableLists refreshRequest={false} />)

    expect(tree).toMatchSnapshot()
  })

  it('should render correctly when no wallet', () => {
    ;(useRecoverableLists as jest.Mock).mockReturnValue({
      wallets: [],
      isFetching: false,
      isRecovering: false,
      startRecovering: jest.fn(),
      recoveringWallet: '',
      modalBackdrop: jest.fn(),
    })
    const tree = renderWithTheme(<RecoverableLists refreshRequest={false} />)

    expect(tree.getAllByText('You are not a guardian of any wallet')).toHaveLength(1)

    expect(tree).toMatchSnapshot()
  })

  it('should call startRecovering', () => {
    const startRecoveringMock = jest.fn()
    ;(useRecoverableLists as jest.Mock).mockReturnValue({
      wallets: ['wallet1'],
      isFetching: false,
      isRecovering: false,
      startRecovering: startRecoveringMock,
      recoveringWallet: '',
      modalBackdrop: jest.fn(),
    })
    const tree = renderWithTheme(<RecoverableLists refreshRequest={false} />)

    fireEvent.press(tree.getByText(' wallet...let1'))

    expect(startRecoveringMock).toHaveBeenCalledWith('wallet1')

    expect(tree).toMatchSnapshot()
  })
})
