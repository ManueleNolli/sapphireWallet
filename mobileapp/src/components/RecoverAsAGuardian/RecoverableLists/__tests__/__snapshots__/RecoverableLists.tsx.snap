// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoverableLists should call startRecovering 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(51, 102, 255, 0.08)",
        "borderColor": "#3366FF",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 13.34,
        "minHeight": 40,
        "minWidth": 40,
        "opacity": 1,
        "paddingHorizontal": 10,
        "paddingVertical": 12,
        "width": "100%",
      }
    }
  >
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#3366FF",
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "bold",
            "marginHorizontal": 10,
          },
        ]
      }
    >
      wallet...let1
    </Text>
  </View>
</View>
`;

exports[`RecoverableLists should render correctly when fetching 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      data-testid="spinner"
    />
  </View>
</View>
`;

exports[`RecoverableLists should render correctly when no wallet 1`] = `
<View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#222B45",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "textAlign": "center",
        },
      ]
    }
  >
    You are not a guardian of any wallet
  </Text>
</View>
`;

exports[`RecoverableLists should render correctly when recovering 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onDismiss={[MockFunction]}
    onRequestClose={[MockFunction]}
    statusBarTranslucent={false}
    transparent={true}
    visible={true}
  >
    <View
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "rgba(143, 155, 179, 0.48)",
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="@backdrop"
      />
      <View
        animationType="fade"
        backdropStyle={
          {
            "backgroundColor": "rgba(143, 155, 179, 0.48)",
          }
        }
        onBackdropPress={[MockFunction]}
        onLayout={[Function]}
        shouldUseContainer={true}
        style={
          [
            undefined,
            {
              "position": "absolute",
            },
            {
              "left": -999,
              "top": -999,
            },
          ]
        }
        visible={true}
      >
        <div
          data-testid="RecoverWallet"
        >
          RecoverWallet
        </div>
      </View>
    </View>
  </Modal>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(51, 102, 255, 0.08)",
        "borderColor": "#3366FF",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 13.34,
        "minHeight": 40,
        "minWidth": 40,
        "opacity": 1,
        "paddingHorizontal": 10,
        "paddingVertical": 12,
        "width": "100%",
      }
    }
  >
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#3366FF",
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "bold",
            "marginHorizontal": 10,
          },
        ]
      }
    >
      wallet...let1
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(51, 102, 255, 0.08)",
        "borderColor": "#3366FF",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 13.34,
        "minHeight": 40,
        "minWidth": 40,
        "opacity": 1,
        "paddingHorizontal": 10,
        "paddingVertical": 12,
        "width": "100%",
      }
    }
  >
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#3366FF",
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "bold",
            "marginHorizontal": 10,
          },
        ]
      }
    >
      wallet...let2
    </Text>
  </View>
</View>
`;
